tools/plugin/modules/ov_noise_suppression/noise_suppression_interface.cpp

```diff
@@ -0,0 +1,176 @@
+		size_t state_size = 0;
+		const char* model_name = std::getenv("NOISE_SUPPRESSION_MODEL_NAME");
+		int i;
+
+		nd = new ns_data();
+		if(!nd)
+			return -ENOMEM;
+
+		*handle = nd;
+		nd->model = core.read_model(model_name);
+		inputs = nd->model->inputs();
+		outputs = nd->model->outputs();
+
+		/* get state name pairs */
+		for (size_t i = 0; i < inputs.size(); i++) {
+			std::string inp_state_name = inputs[i].get_any_name();
+			if (inp_state_name.find("inp_state_") == std::string::npos)
+			    continue;
+
+			std::string out_state_name(inp_state_name);
+			out_state_name.replace(0, 3, "out");
+
+			/* find corresponding output state */
+			if (outputs.end() == std::find_if(outputs.begin(), outputs.end(), [&out_state_name](const ov::Output<ov::Node>& output) {
+				return output.get_any_name() == out_state_name;
+			}))
+				return -EINVAL;
+
+			nd->state_names.emplace_back(inp_state_name, out_state_name);
+
+			ov::Shape shape = inputs[i].get_shape();
+			size_t tensor_size = std::accumulate(shape.begin(), shape.end(), 1,
+							     std::multiplies<size_t>());
+
+			state_size += tensor_size;
+		}
+
+		if (!state_size)
+			return -EINVAL;
+
+		/*
+		 * compile the model for the CPU and save the infer_request objects for each
+		 * channel separately
+		 */
+		ov::CompiledModel compiled_model = core.compile_model(nd->model, "CPU", {});
+		for (i = 0; i < NS_MAX_SOURCE_CHANNELS; i++)
+			nd->infer_request[i] = compiled_model.create_infer_request();
+
+		nd->inp_shape = nd->model->input("input").get_shape();
+
+		return 0;
+	}
+
+	void ov_ns_free(ns_handle handle) {
+		struct ns_data *nd = (struct ns_data *)handle;
+
+		delete nd;
+	}
+
+	int ov_ns_process(ns_handle handle,
+			  struct input_stream_buffer *input_buffers, int num_input_buffers,
+			  struct output_stream_buffer *output_buffers, int num_output_buffers)
+	{
+		struct audio_stream *source = (struct audio_stream *)input_buffers[0].data;
+		struct audio_stream *sink = (struct audio_stream *)output_buffers[0].data;
+		struct ns_data *nd = (struct ns_data *)handle;
+		std::vector<float> inp_wave_fp32, out_wave_fp32;
+		/* only 16-bit supported for now */
+		int16_t *input_data = (int16_t *)audio_stream_get_rptr(source);
+		int16_t *output_data = (int16_t *)audio_stream_get_wptr(sink);
+		uint32_t frame_count = input_buffers[0].size;
+		float scale = 1.0f / std::numeric_limits<int16_t>::max();
+		int i, j, ch;
+
+		/*
+		 * The noise suppression model only supports mono, so process each channel
+		 * separately
+		 */
+		inp_wave_fp32.resize(frame_count, 0);
+		out_wave_fp32.resize(frame_count, 0);
+
+		for (ch = 0; ch < NS_MAX_SOURCE_CHANNELS; ch++) {
+			/* split each channel samples and convert to floating point */
+			for (i = ch, j = 0; j < frame_count; i+=2,j++) {
+				void *inp = &input_data[i];
+
+				/* wrap if needed */
+				if (inp >= source->end_addr)
+			                inp = (char *)source->addr +
+			                        ((char *)inp - (char *)source->end_addr);
```

Comment:
This works but it's not efficient. The linear buffer can be filled in two steps with help from audio stream functions for samples or frames before wrap. We've fixed a lot of these during last year for better performance.

Reply:
@singalsu lets take this as an incremental fix. Can you give a pointer to good example code to copy.

Reply:
Yes it's ok, this is not as consuming as read/write frag.